Support up to hhvm 4.25

lexidor opened the pull request:
Runs many tests in paralel

lexidor committed:
Multiplex tests

lexidor committed:
Use php AND hhvm to composer install

lexidor committed:
Correct runtime

lexidor committed:
Ignore all dependencies

Just use the newest thing that works with your hhvm version

lexidor committed:
Stop locking dependencies

This is not practical when building against x hhvm versions

lexidor committed:
Fix return type inferance on lambda's

lexidor committed:
Ignore references during linting

lexidor committed:
Use safe version constraints

lexidor committed:
Indentation error

lexidor committed:
Indent error

lexidor committed:
Test on more versions

lexidor commented:
I'll need to see if we can reduce the version requirements of hsl and hsl-experimental.

lexidor committed:
Reduce version requirements

lexidor committed:
Use hhvm 3.28.3, since there is no latest

lexidor committed:
Remove hsl-experimental

There are some issues in hhvm 3.28, but this _might_ very well fix that.

tyronjung-quizlet commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 vendor
 **/*.hhast.parser-cache
+composer.lock
```
```suggestion
composer.lock

```
Though, don't we want to track `composer.lock`?

lexidor replied:
I am afraid you can't do that.
Take a look at the other Hack libraries
They all git ignore the lock file.
Travis should not be able to see it, otherwise the tests fail.
Composer install would install the old dependencies in newer versions of hhvm.

tyronjung-quizlet commented on .travis.sh:
```diff
@@ -9,7 +9,13 @@ php --version
 )
-hhvm /usr/local/bin/composer install
+
+runtime=$(hhvm --php -r "echo HHVM_VERSION[0] === '3' ? 'hhvm' : 'php';")
```
Should this be:
```suggestion
runtime=$(hhvm --php -r "echo HHVM_VERSION[0] === '3' ? 'php' : 'hhvm';")
```
?

lexidor replied:
No, this is the right way around.
If your hhvm version starts with a 3,
hhvm should be in charge of running composer.
If you are hhvm 4 or up,
let php do it.
Hhvm 4 can't run composer.

tyronjung-quizlet replied:
Gotcha, thanks.

tyronjung-quizlet commented on .travis.yml:
```diff
@@ -4,7 +4,19 @@ os:
 language: shell
 services: docker
 env:
+- HHVM_VERSION=3.28.3
```
Should we start from `3.25.3` like in [fb_intercept_polyfill](https://github.com/lexidor/fb_intercept_polyfill/blob/ef93b77e4753feefc77206ec9a954fde4c318929/.travis.yml#L5)?

lexidor replied:
Yes, but that requires making the same changes to the Travis dot yaml as fb_intercept_polyfill.
I did have this done back when I wrote this. 
I'll see if it can be copy pasted verbatim.

tyronjung-quizlet commented on hhast-lint.json:
```diff
@@ -6,8 +6,9 @@
     }
   ]
-}
+}
```
```suggestion
}

```

lexidor committed:
Update .git ignore with a trailing newline

Co-Authored-By: Tyron Jung <41349925+tyronjung-quizlet@users.noreply.github.com>

lexidor committed:
Update hhast-lint.json with a trailing newline

Co-Authored-By: Tyron Jung <41349925+tyronjung-quizlet@users.noreply.github.com>

lexidor committed:
Use the same version detection as fb_intercept

lexidor committed:
Test on older hhvm versions

lexidor committed:
Remove the hh-clilib dependency

It is causing composer install problems on older versions.
It is also unused as far as I can tell.
Grepping for Facebook or CLILib finds no hits.

lexidor committed:
Add hh_fixme and comment on why

lexidor committed:
Remove some hh_fixmes

lexidor committed:
Readd tests for newer hhvm versions

lexidor committed:
Update a comment that counted the number of fixmes

tyronjung-quizlet commented on .travis.sh:
```diff
@@ -9,10 +9,28 @@ php --version
+  hhvm /usr/local/bin/composer install
+else
+    # Implicitly uses php
+    composer install
```
Could we get consistent indentation for this `else` block?

tyronjung-quizlet commented on composer.json:
```diff
@@ -13,14 +13,13 @@
-        "facebook/hh-clilib": "^1.2.1|^2"
+		  	"hhvm": ">=3.24",
+		  	"hhvm/hhvm-autoload": "^1|^2|^3",
+		  	"hhvm/hsl": ">1.4 <5"
```
And consistent indentation here?

lexidor replied:
Yeah, this is a classic case of _it looks good with my tab width._
I switch tab widths back and forth all the time, since 2 is the Hack default,
but 4 is what we use for our stuff.
This is something I overlooked.
Thanks for telling me about how bad these files looked.

tyronjung-quizlet commented on src/MockManager.php:
```diff
@@ -325,10 +325,17 @@ protected static function getValidatedObjectMethodMockKeyAndObjectHash<T>(
 	}
 
+	/**
+	 * @return void This function doesn't return anything.
```
```suggestion
	 * @return mixed This function doesn't actually return anything.
```

tyronjung-quizlet reviewed (approved):
Thanks for making Hammock work for HHVM 3 _and_ 4. 😄
Left a few minor comments, but they're non-blocking.

lexidor committed:
Be correct about the return type of mock()

I used phpdoc to indicate that this function returns void,
but it is best to not contradict the hack typehint.
The comment itself explains this perfectly well.

Co-Authored-By: Tyron Jung <41349925+tyronjung-quizlet@users.noreply.github.com>

lexidor committed:
Whitespace changes

tyronjung-quizlet reviewed (approved):
👍

lexidor commented:
:laughing: One downside of supporting more hhvm versions is that Travis will become a lot slower.
I _randomly_ decided to pick one release for each month.
If the slow Travis times bother you or reduce your productivity, feel free to change it up.

tyronjung-quizlet commented:
> If the slow Travis times bother you or reduce your productivity, feel free to change it up.

Will do. ~Feel free to merge when ready.~ Will merge on your behalf.
